# Сбор данных кинопоиск и внесение в базу данных neo4j (python)

### Импортируем необходимые библиотеки

In [1]:
import requests
import pandas as pd 
from py2neo import Graph, Node, Relationship

Библиотеки, которые будем использовать: 
* requests - библиотека для запроса и получения данных с сайта 
* pandas - библиотека для создания таблицы из спарсинных данных
* py2neo - предназначена для работы с графовой базой данных Neo4j в Python

### Создаем функцию по сбору данных и сохранения в dataframe pandas

In [2]:
def fetch_movie_data(start_FilmId, end_FildId):
    # Задаем нужные url и доступ к API
    url1 = 'https://kinopoiskapiunofficial.tech/api/v2.2/films/' # URL для получения данных о фильме
    url2 = 'https://kinopoiskapiunofficial.tech/api/v1/staff?filmId=' # URL для получения данных об актерах фильма
    headers = {
    'X-API-KEY': '22deee09-c581-4c94-ab8e-2ad3602a5679',
    'Content-Type': 'application/json'
    }
    final_dataframe = pd.DataFrame()
    # Запускаем цикл сбора данных с заданных FilmId
    for i in range(start_FilmId, end_FildId):
        response1 = requests.get(url=f'{url1}{i}', headers=headers)
        response2 = requests.get(url=f'{url2}{i}', headers=headers)
        # Проверяем существует ли такая страница и начинаем преобразовывать данные
        if response1.status_code == 200 and response2.status_code == 200:
            # Собираем данные о фильме
            json_response1 = response1.json()
            data1 = pd.DataFrame([json_response1])
            data1_v = data1[['nameRu', 'year', 'slogan']]
            # Собираем данные об участниках фильма
            json_response2 = response2.json()
            data2 = pd.DataFrame(json_response2)
            df3 = data2[['nameRu', 'professionKey']]
            new_df = df3.groupby('professionKey').agg({'nameRu': list}).reset_index()
            new_df4 = new_df.set_index('professionKey')
            new_df5 = new_df4.transpose()
            new_df6 = new_df5.reset_index(drop=True)
            # Уберем пустые ('') строки, чтобы не создавать лишних пустых узлов в neo4j
            new_df6['ACTOR'] = new_df6['ACTOR'].apply(lambda x: [item for item in x if item != ''])
            new_df6['DIRECTOR'] = new_df6['DIRECTOR'].apply(lambda x: [item for item in x if item != ''])
            new_df6['PRODUCER'] = new_df6['PRODUCER'].apply(lambda x: [item for item in x if item != ''])
            new_df6['WRITER'] = new_df6['WRITER'].apply(lambda x: [item for item in x if item != ''])
            combined_df = pd.concat([data1_v, new_df6], axis=1)
            # Соединяем две таблицы
            final_dataframe = pd.concat([final_dataframe, combined_df], ignore_index=True)
        else:
            print(f"Ошибка получения данных для фильма с ID {i}: {response1.status_code}")
    return final_dataframe

# Задаем нужные id фильмов для сбора информации
start_FilmId = 300
end_FildId = 320
# Сохраняем полученные данные в переменную result_dataframe
result_dataframe = fetch_movie_data(start_FilmId, end_FildId)

Ошибка получения данных для фильма с ID 315: 404


Здесь взяты 19 фильмов с id 300-319, если нужно больше, то ставить необходимо другие рамки. 

***Ошибка вышла с id 315 - так как такого фильма с таким id не существует (ошибка 404)***

### Теперь покажем какая получилась таблица 

In [3]:
result_dataframe

nameRu  year  \
0                                       Лиззи Магуайр  2003   
1                                             Матрица  1999   
2                                                Клад  2003   
3                                   Управление гневом  2003   
4                                               Афера  2003   
5                                    Телефонная будка  2002   
6                                              Рекрут  2003   
7                              Разыскиваются в Малибу  2003   
8   Тупой и еще тупее тупого: Когда Гарри встретил...  2003   
9                                         Эквилибриум  2002   
10                                      Дежурный папа  2003   
11                                  Пуленепробиваемый  2003   
12                      Властелин колец: Две крепости  2002   
13                                  Играй, как Бекхэм  2002   
14                                            Цыпочка  2002   
15                                 Матрица: Революция  2003   
16                              Любовь с уведомлением  2002   
17                           Ограбление по-итальянски  2003   
18                      Терминатор 3: Восстание машин  2003   

                                               slogan  \
0                         Lizze McGuire goes to Rome.   
1                     Добро пожаловать в реальный мир   
2                         Кто ищет, тот всегда найдёт   
3                               Let the healing begin   
4                Это не про деньги, это про ДЕНЬГИ!!!   
5                           Линия жизни на линии огня   
6   Trust. Betrayal. Deception. In the C.I.A. noth...   
7                    Ever feel like you don't belong?   
8                             Тупой с первого взгляда   
9                    Мир, где эмоции караются смертью   
10                                    D-Day Is Coming   
11  To fulfill the prophecy, he must find the one ...   
12                           Приключение продолжается   
13                   Живи на полную! Люби без памяти!   
14  Однажды она проснулась не в себе... В буквальн...   
15               Все, что имеет начало, имеет и конец   
16       Он стремился к своей мечте, а она была рядом   
17                            Хватай. Убегай. Отомсти   
18                             The Machines Will Rise   

                                                ACTOR  \
0   [Хилари Дафф, Адам Лэмберг, Халли Морган, Робе...   
1   [Киану Ривз, Лоренс Фишбёрн, Кэрри-Энн Мосс, Х...   
2   [Сигурни Уивер, Джон Войт, Патриша Аркетт, Шай...   
3   [Адам Сэндлер, Джек Николсон, Мариса Томей, Лу...   
4   [Эдвард Бёрнс, Рэйчел Вайс, Моррис Честнат, Ли...   
5   [Колин Фаррелл, Кифер Сазерленд, Форест Уитаке...   
6   [Аль Пачино, Колин Фаррелл, Бриджет Мойнэхэн, ...   
7   [Джейми Кеннеди, Тэй Диггз, Энтони Андерсон, Р...   
8   [Эрик Кристиан Олсен, Дерек Ричардсон, Шери От...   
9   [Кристиан Бэйл, Тэй Диггз, Энгус Макфадьен, Шо...   
10  [Эдди Мерфи, Стив Зан, Джефф Гарлин, Реджина К...   
11  [Чоу Юнь-Фат, Шонн Уильям Скотт, Джейми Кинг, ...   
12  [Элайджа Вуд, Иэн Маккеллен, Вигго Мортенсен, ...   
13  [Парминдер К. Награ, Кира Найтли, Джонатан Риз...   
14  [Роб Шнайдер, Анна Фэрис, Мэттью Лоуренс, Эрик...   
15  [Киану Ривз, Хьюго Уивинг, Кэрри-Энн Мосс, Лор...   
16  [Сандра Буллок, Хью Грант, Алисия Уитт, Дэна А...   
17  [Шарлиз Терон, Марк Уолберг, Дональд Сазерленд...   
18  [Арнольд Шварценеггер, Ник Стал, Клэр Дэйнс, К...   

                   COMPOSER  \
0         [Клифф Эйдельман]   
1               [Дон Дэвис]   
2           [Джоэл МакНили]   
3        [Тедди Кастелуччи]   
4             [Кристоф Бек]   
5   [Гарри Грегсон-Уильямс]   
6           [Клаус Бадельт]   
7      [Джон  Ван Тонгерен]   
8            [Эбан Шлеттер]   
9           [Клаус Бадельт]   
10           [Дэвид Ньюман]   
11             [Эрик Серра]   
12             [Говард Шор]   
13            [Крэйг Прюсс]   
14             [Джон Дебни]   
15      

### Теперь подключаемся к базе данных neo4j и заносим необходимую информацию

In [4]:
# Подключение к базе данных Neo4j
graph = Graph("bolt://localhost:7687", auth=("neo4j", "password"))

# Создаем словарь действия в фильме каждого участника
profession_types = {
    'ACTOR' : 'ACTED_IN',
    'DIRECTOR' : 'DIRECTED',
    'PRODUCER' : 'PRODUCED',
    'WRITER' : 'WROTE'
}

# Запускаем цикл создания узлов фильмов и участников в neo4j
for index, record in result_dataframe.iterrows():
    movie_title = record['nameRu']
    # Создание узла для фильма
    movie_node = Node("Movie", title=movie_title, year=record['year'], slogan=record['slogan'])
    graph.merge(movie_node, "Movie", "title")
    # Создание узлов для участников
    for profession in ['ACTOR', 'DIRECTOR', 'PRODUCER', 'WRITER']:
        if profession in record:
            for person_name in record[profession]:
                person_node = Node('Person', name=person_name)
                graph.merge(person_node, 'Person', 'name')
                relationship_type = Relationship(person_node, profession_types[profession], movie_node)
                graph.merge(relationship_type) 